### PR TITLE
Reduced the frequency of 'server failed to send complete response' exceptions when using S3AsyncClient.

### PR DESCRIPTION
**Pull request by millems**
Should help reduce #1122 with S3.

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java [hunk 1]
```diff
@@ -121,21 +121,26 @@ public void onNext(ByteBuffer byteBuffer) {
...
-                            } else {
-                                subscription.request(1);
-                            }
-                            writeInProgress = false;
+                    position.addAndGet(result);
+                    synchronized (FileSubscriber.this) {
+                        if (closeOnLastWrite) {
+                            close();
+                        } else {
+                            subscription.request(1);
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java]: We don't seem to check that the buffer was written entirely before requesting another one. Seems like an edge case (fine with addressing in a different PR).

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java [hunk 2]
```diff
@@ -121,21 +121,26 @@ public void onNext(ByteBuffer byteBuffer) {
             if (byteBuffer == null) {
                 throw new NullPointerException("Element must not be null");
             }
-            
+
+            if (byteBuffer.remaining() == 0) {
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java]: Is this just an optimization? The docs say the write finishes immediately if the buffer is empty

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java (hunk 1 above)
Comment (millems) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java]: Do you know what the expectation is when the buffer was not written entirely?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java (hunk 2 above)
Comment (millems) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java]: I can update it, then. Missed that part of the docs.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java (hunk 1 above)
Comment (millems) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java]: Fixed.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java (hunk 2 above)
Comment (millems) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java]: Fixed.

**Pull request merged**